include/libdnf/repo/repo_cache.hpp

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_REPO_REPO_CACHE_HPP
+#define LIBDNF_REPO_REPO_CACHE_HPP
+
+#include "libdnf/base/base_weak.hpp"
+#include "libdnf/common/exception.hpp"
+
+#include <filesystem>
+#include <string>
+
+
+namespace libdnf::repo {
+
+
+struct RepoCacheRemoveStatistics {
+    std::size_t files_removed;  // Number of removed files and links.
+    std::size_t dirs_removed;   // Number of removed directorires.
+    std::size_t errors;         // Numbes of errors.
+
+    RepoCacheRemoveStatistics & operator+=(const RepoCacheRemoveStatistics & rhs) noexcept;
+};
+
+
+/// RepoCache exception
+class RepoCacheError : public Error {
+public:
+    using Error::Error;
+    const char * get_domain_name() const noexcept override { return "libdnf::repo"; }
+    const char * get_name() const noexcept override { return "RepoCacheError"; }
+};
+
+
+/// Repository cache management class.
+class RepoCache {
+public:
+    using RemoveStatistics = RepoCacheRemoveStatistics;
+
+    /// The name of the attribute used to mark the cache as expired.
+    static constexpr const char * ATTRIBUTE_EXPIRED = "expired";
+
+    /// Construct a new repository cache management instance.
+    ///
+    /// @param base            WeakPtr on the Base instance.
+    /// @param repo_cache_dir  Path to repository cache directory.
+    RepoCache(const libdnf::BaseWeakPtr & base, const std::filesystem::path & repo_cache_dir);
+
+    /// Construct a new repository cache management instance.
+    ///
+    /// @param base            Base instance reference.
+    /// @param repo_cache_dir  Path to repository cache directory.
+    RepoCache(libdnf::Base & base, const std::string & repo_cache_dir);
```

Comment:
I am just curious whether we need public constructors?

Reply:
May be it will sound strange but why we cannot use a repo object in constructor to replace base and path. Then it can take the path directly from `repo` config. What do you think?

Reply:
> but why we cannot use a repo object in constructor to replace base and path

This would mean that I would have to create repo objects for all caches in the `clean` command implementation.

> Then it can take the path directly from repo config.

The storage configuration may no longer exist. Imagine that a user deletes a repository configuration and later calls the `clean` command.

Reply:
Thank you for explanation.